### ISSUE TITLE
C#10 Improved Definite Assignment Analysis null-conditional operator change not properly version-gated

### ISSUE THREAD

**Issue by khyperia** [dotnet/roslyn#60084]
First off, I just want to say that I have no idea what I'm doing. It's really difficult to do research here on what's a bug, what's intended behavior, and when things changed. So, my apologies if this issue isn't actually a bug after all!

---

Repro steps:

`ConsoleApp1.csproj`:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <LangVersion>6.0</LangVersion>
  </PropertyGroup>
</Project>
```

`Program.cs`:
```cs
class Testing {
    private bool Test(out object thing) {
        thing = new object();
        return true;
    }
    private static void Thingy(Testing testing) {
        object obj;
        if (testing?.Test(out obj) == true) {
            obj.ToString();
        }
    }
    private static void Main() => Thingy(new Testing());
}
```

I _believe_ that this **should fail** compilation, with the error `error CS0165: Use of unassigned local variable 'obj'`, because LangVersion is 6.0. For example, this was reported as happening in https://github.com/dotnet/roslyn/issues/45776, and the response was "this is intended behavior, file a language change if you want to change the behavior", which I believe eventually turned into the C#10 Improved Definite Assignment Analysis feature.

However, compilation **succeeds** using visual studio `17.0.5`, C# tools `4.0.1-1.21568.1+6ab6601178d9fba8c680b56934cd1742e0816bff`, even though LangVersion is set to 6.0. (I cycled down through language versions from 9.0 to 6.0, and stopped at 6.0 because that's when `?.` was introduced. All succeeded compilation.)

When using an **old** compiler, set to language version 9.0 - e.g. how I discovered this, the one in Unity 2021.2.7f1, or the one currently [online on sharplab](https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgAQBUCmBnYYAdgOboDeAsAFDq3oAOATmAG4CGw26ARgPa8QseYAApeAV2Dpe3AFbYAxlOAALIsQCU5anV3pV69AF50hbAHdpcxaI0BuHXtrIA7PsbjsDmnQC+j2iZWDi5kAEYANnRkABYsNRIATxEcfENONJItSh8nGXklK1lvJ1owADN0EQyCEgB+ADpU0QkpfK0jE2APbGyA0tp8pt4AZW71EXt+3X9c9FndIPZOaMjouIBZNiJJ4wA+ePVks0tmiY0pqlmgA=), compilation **fails** (as it should).

Thanks for your time!

---

Edit: ah, sorry, just realized my VS was of date, updated to `17.1.1`, C# tools `4.1.0-5.22109.6+0c82c4114a4e4b8b723b915eee3b13261db6717f`, but compilation still incorrectly succeeds.

**Comment by RikkiGibson** [comment]
Thanks for filing this issue. We intentionally shipped the behavior you described.

To give a LangVersion error here we end up needing to answer the question: *would definite assignment have given an error if the "improved" rules weren't present*? We think this would require either:

1. Introducing a new flow state to definite assignment, e.g. "DefinitelyAssignedInCSharp10".
2. Adding a flag that lets us run with either the "old" or "new" rules, then in pre-c# 10, running first with the "old" rules, and if errors are present, running it again to decide if it would have succeeded with the new rules, so that we can give a LangVersion diagnostic.

We weren't happy with the cost of either of these so we decided to just behave as though the feature is always available, even though it will fail in older compilers.

No need to apologize, it's on us that we didn't make this decision more discoverable. 🙂

**Comment by khyperia** [comment]
Ah, that makes sense, thank you!
